### DISCUSSION TITLE
Saving and/or importing nonlinear warps

### DISCUSSION THREAD

**Discussion by KevinAquino** [dipy/dipy#2359] [Registration] [answered]
## Description
I was wondering if there are any helper functions that exist to save nonlinear warps (specifically with SymmetricDiffeomorphicRegistration) and are there any helper functions that allow one to import warps from FSLs' FNIRT algorithm. 

I have managed to go through DiPy to find a way to apply a warping field, but FSL seems to store co-efficients of the polynomials that make the up the warping fields. Without having to re-invent the wheel, are there any helper tools that do this?

Thank you! 
Also DiPY is so great - I'm also incredibly impressed how fast SymmetricDiffeomorphicRegistration runs - I can seem to get an incredible warp in minutes vs hours in ANTs.

**Comment by skoudoro** [comment, marked as answer]
Hi @KevinAquino,

> if there are any helper functions that exist to save nonlinear warps

As you can see [here](https://github.com/dipy/dipy/blob/master/dipy/align/_public.py#L218-L281), we use:
```python
from dipy.align import write_mapping, read_mapping
```
they are saved as Nifti files.

> there any helper functions that allow one to import warps from FSLs' FNIRT algorithm.

Not yet, and it would be a great contribution! this is on a TODO list but feel free to contribute if you create one helper function.

> Also DiPY is so great - I'm also incredibly impressed how fast SymmetricDiffeomorphicRegistration runs - I can seem to get an incredible warp in minutes vs hours in ANTs.

Great !  there is still some work going on to improve performance and the overall design of our registration framework, but it is tricky to balance readability and performance. 

Thank you for your feedback

**Reply by Lihi-Gur-Arie** [reply]
Thanks! 
I saved with dipy.align.write_mapping  as you suggested, thoug I'm not sure how to open it. 
The read_mapping function expects expects takes 3 arguments: disp, domain_img and codomain_img. What are the last two?
(I made SymmetricDiffeomorphicRegistration on a single image)

**Reply by arokem** [reply]
Hi Lihi, 

Sorry - I agree that this terminology is a bit opaque. 

The pattern we use is shown in [this test](https://github.com/dipy/dipy/blob/61a6f14015b8cd58b48501e1ef6fc853e55b6807/dipy/align/tests/test_api.py#L49-L67). 

To simplify, imagine that you have two nibabel `Nifti1Image` objects that you are registering to each other with syn (registering the "moving" image to the "static" image), this is the pattern you would to write:
```
warped_moving, mapping = syn_registration(moving, static)
write_mapping(mapping, 'mymapping.nii.gz')
```

Which you can then read as:
```
mapping = read_mapping('mymapping.nii.gz', moving, static)
```

Does that work for your use-case?

**Comment by arokem** [comment]
> > there any helper functions that allow one to import warps from FSLs' FNIRT algorithm.

> Not yet, and it would be a great contribution! this is on a TODO list but feel free to contribute if you create one helper function.

For now, you might want to check out [fslpy](https://pypi.org/project/fslpy/)

**Comment by KevinAquino** [comment]
Great thank you both for that!

I didn't know about fslpy thank you @arokem for that! 

Regarding a helper function @skoudoro if i create one ill be sure to send a PR! :) thank you!
